### PR TITLE
Introduce `ResourceAdapterLifecycleListener`

### LINKED ISSUE #103: To pause or defer the listening of messages by a JMS resource adapter
My Quarkus application uses `quarkus-ironjacamar` extension to integrate message consumption from an Artemis JMS queue.

This application defines a ResourceAdapterFactory subclass which one finally wraps to a JMSMessageEndpoint (as MessageEndpoint) from a MessageEndpoint and a MessageListener.

```
@ResourceAdapterKind(value = "artemis")
@ResourceAdapterTypes(connectionFactoryTypes = { jakarta.jms.ConnectionFactory.class, XAConnectionFactory.class })
public class ArtemisResourceAdapterFactory implements ResourceAdapterFactory {
  ...
  @Override
  public MessageEndpoint wrap(MessageEndpoint endpoint, Object resourceEndpoint) {
      return new JMSMessageEndpoint(endpoint, (MessageListener) resourceEndpoint);
  }
  
  private static class JMSMessageEndpoint extends MessageEndpointWrapper implements MessageListener {
  ...
      @Override
      public void onMessage(Message message) {
          listener.onMessage(message);
      }
  }
}
```

I need to pause or delay the message listening in order to perform some task before any message consumption, i.e. loading a cache.

### PR DESCRIPTION
**Pull request by gastaldi**
This should allow to register pre and post-callbacks when a ResourceAdapter starts and stops.

- Fixes #103

### CODE REVIEW

**Comment by gastaldi** [comment]
@vondacho can you give this PR a try and see if that works for you?

File: deployment/src/main/java/io/quarkiverse/ironjacamar/deployment/IronJacamarProcessor.java [hunk 1]
```diff
@@ -265,17 +281,26 @@ void initDefaultBootstrapContext(BeanContainerBuildItem beanContainerBuildItem,
...
+            ApplicationIndexBuildItem applicationIndex,
             List<ContainerCreatedBuildItem> containers,
             IronJacamarRecorder recorder,
             CoreVertxBuildItem vertxBuildItem,
             BeanContainerBuildItem beanContainerBuildItem,
             BuildProducer<ContainerStartedBuildItem> startedProducer) {
+        List<Class<ResourceAdapterLifecycleListener>> listeners = new ArrayList<>();
+        for (ClassInfo type : applicationIndex.getIndex().getAllKnownImplementors(ResourceAdapterLifecycleListener.class)) {
+            try {
+                listeners.add((Class<ResourceAdapterLifecycleListener>) Class.forName(type.name().toString()));
```
Comment (gastaldi) [inline deployment/src/main/java/io/quarkiverse/ironjacamar/deployment/IronJacamarProcessor.java:293]:
```suggestion
                listeners.add((Class<ResourceAdapterLifecycleListener>) Class.forName(type.name().toString(), true, Thread.currentThread().getContextClassloader()));
```

**Comment by gastaldi** [comment]
@vondacho I've updated the PR, can you give it another try please?

File: deployment/src/main/java/io/quarkiverse/ironjacamar/deployment/IronJacamarProcessor.java [hunk 2]
```diff
@@ -265,17 +281,29 @@ void initDefaultBootstrapContext(BeanContainerBuildItem beanContainerBuildItem,
...
+                .getAllKnownImplementors(ResourceAdapterLifecycleListener.class);
+        List<Class<ResourceAdapterLifecycleListener>> listeners = new ArrayList<>(allKnownImplementors.size());
+        for (ClassInfo type : allKnownImplementors) {
+            try {
+                listeners.add((Class<ResourceAdapterLifecycleListener>) Class.forName(type.name().toString(), true,
+                        Thread.currentThread().getContextClassLoader()));
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (geoand) [reply deployment/src/main/java/io/quarkiverse/ironjacamar/deployment/IronJacamarProcessor.java:300]: Do you really need the classes themselves?

[the remark this answers is not in the document]
Comment (gastaldi) [reply deployment/src/main/java/io/quarkiverse/ironjacamar/deployment/IronJacamarProcessor.java:300]: Well, I call BeanContainer.beanInstance in the `IronJacamarVerticle`, so I think so?

[the remark this answers is not in the document]
Comment (geoand) [reply deployment/src/main/java/io/quarkiverse/ironjacamar/deployment/IronJacamarProcessor.java:300]: You can do this yeah, but if I were you I would just pass the list of Strings representing the class names and load the classes in the recorder

**Review by vondacho — APPROVED** [review]
Thanks. Seems good to me.

**Pull request merged**